File: constraint-union-vectors-run.py
#!/usr/bin/env python3
"""Run the CLC-v1 §7.1 ConstraintUnion vectors (rev CLC-1.12)."""

import json
import os
import sys

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
from clc_semantics import constraint_union, CLCError  # noqa: E402


def main():
    path = os.environ.get("CLC_CONSTRAINT_UNION_VECTORS")
    if not path:
        here = os.path.dirname(os.path.abspath(__file__))
        path = os.path.join(here, "..", "capability", "data", "_vectors", "clc-v1", "constraint-union-vectors.json")
    with open(path) as f:
        vectors = json.load(f)

    passed = failed = 0
    for v in vectors:
        want_reason = v["expect"].get("reason")
        try:
            got = constraint_union(v["chain"])
            err = None
        except CLCError as e:
            got = None
            err = str(e)
        if want_reason:
            ok = err is not None and err.split(":", 1)[0] == want_reason
        else:
            ok = err is None and got == v["expect"]["union"]
        if ok:
            passed += 1
        else:
            failed += 1
            print("%-8s FAIL want=%s/%s got=%s err=%s" % (
                v["id"], v["expect"].get("union"), want_reason, got, err))
    print("Total: %d | Pass: %d | Fail: %d" % (len(vectors), passed, failed))
    if failed:
        sys.exit(1)


if __name__ == "__main__":
    main()

File: ts/constraint-union-vectors-run.ts
// CLC-v1 §7.1 ConstraintUnion vectors runner (rev CLC-1.12).
import { readFileSync } from 'node:fs';
import { dirname, join, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';
import { constraintUnion } from './clc_semantics.ts';
import type { Grant } from './clc_semantics.ts';

interface Expectation { union?: string[]; reason?: string }
interface Vector { id: string; chain: Grant[]; expect: Expectation }
const here = dirname(fileURLToPath(import.meta.url));
const path = process.env.CLC_CONSTRAINT_UNION_VECTORS
    || resolve(join(here, '..', '..', 'capability', 'data', '_vectors', 'clc-v1', 'constraint-union-vectors.json'));
const vectors: Vector[] = JSON.parse(readFileSync(path, 'utf8'));
let pass = 0;
let fail = 0;
for (const v of vectors) {
    let got: string[] | null = null;
    let err: string | null = null;
    try {
        got = constraintUnion(v.chain);
    } catch (e) {
        err = (e as Error).message;
    }
    let ok: boolean;
    if (v.expect.reason) {
        ok = err !== null && err.split(':', 1)[0] === v.expect.reason;
    } else {
        const want = v.expect.union ?? [];
        ok = err === null && got !== null && got.length === want.length && got.every((x, i) => x === want[i]);
    }
    if (ok) pass++;
    else { fail++; console.log(`${v.id} FAIL want=${JSON.stringify(v.expect)} got=${JSON.stringify(got)} err=${err}`); }
}
console.log(`Total: ${vectors.length} | Pass: ${pass} | Fail: ${fail}`);
if (fail) process.exit(1);
